### PR TITLE
CylindricalEndcap: Allow target sphere to be inside source sphere.

### LINKED ISSUE #2879: Random failure in Unit.Domain.CoordinateMaps.CylindricalEndcap
# Bug reports:

### Expected behavior:

Test doesn't fail

### Current behavior:

Test fails with 
```
/__w/spectre/spectre/tests/Unit/Helpers/Domain/CoordinateMaps/TestMapHelpers.hpp:174: FAILED:
  CHECK( gsl::at(gsl::at(expected_identity, i), j) == approx(i == j ? 1. : 0.) )
with expansion:
  -0.0000000000000302 == Approx( 0.0 )
with messages:
  CylindricalEndcap
  Seed is: 2824750152 from /__w/spectre/spectre/tests/Unit/Domain/
  CoordinateMaps/Test_CylindricalEndcap.cpp:21
  center_one := { -0.85538704758621464, -0.649336057119412, 0.35899907476230597
  }
  center_two := { 0.82687725167985127, -0.39149005589612873, 0.
  08637647582608521 }
  radius_one := 0.83982241234364063
  z_plane := 0.46899799388377666
  radius_two := 3.82498471745980595
  proj_center := { -1.79840685186027738, -0.74104139314354389, -2.
  32206165570202927 }
  Seed is: 3392210867 from /__w/spectre/spectre/tests/Unit/Helpers/Domain/
  CoordinateMaps/TestMapHelpers.hpp:532
  height := -0.76291987755007407
  phi := 3.06171112463691575
  radius := 0.47393096255658523
```

### Environment:

This failed on CI on my fork.  If the link doesn't work, let me know and I will add more details...

https://github.com/kidder/spectre/runs/1882348286?check_suite_focus=true

Add as an attachment `$SPECTRE_BUILD_DIR/LibraryVersions.txt` or
add its contents here.

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

CylindricalEndcap is a map that maps a 3d unit right cylinder to the portion of a volume contained between two spheres, where the volume is bounded by the intersection of the 'source' sphere with a z=constant plane.
For convenience and simplicity, CylindricalEndcap demands that the 'target' sphere contains the 'source' sphere.
However, it turns out that for the cylindrical BBH domain we also need CylindricalEndcap for the case in which the 'source' sphere contains the 'target' sphere.

This PR modifies CylindricalEndcap so that it allows the case in which the 'source' sphere contains the 'target' sphere.
The changes necessary were:
  - Change the logic choosing a particular root of a quadratic equation in the formula for the map and its inverse; the correct root to choose depends on the relationship between the two spheres.
  - Change the sanity checks in the CylindricalEndcap constructor, which now allow either sphere to be contained in the other.
  - Add the newly allowed configurations to Test_CylindricalEndcap.cpp.
  - Simplify the existing test in Test_CylindricalEndcap.cpp and make it (and the matching sanity checks in CylindricalEndcap.cpp) more robust.
  - Update the documentation.
  
 This should fix #2879

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by markscheel** [comment]
Rebased onto develop and fixed many conflicts.

**Comment by markscheel** [comment]
Updated the last commit to fix a gcc compiler warning.

File: src/Domain/CoordinateMaps/FocallyLiftedMap.hpp [hunk 1]
```diff
@@ -99,6 +99,17 @@ bool operator==(const FocallyLiftedMap<InnerMap>& lhs,
  * and we demand that the \f$\sigma=0\f$ and \f$\sigma=1\f$ surfaces do
  * not intersect; otherwise the map is singular.
  *
+ * Also note that the quadratic Eq. (2) typically has more than one
+ * root, corresponding to two intersections of the sphere.  The
+ * boolean parameter `source_is_between_focus_and_target` that is
+ * passed into the constructor of `FocallyLiftedMap` is used to choose the
+ * appropriate root, or to die with an error messages if a suitable
```
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/FocallyLiftedMap.hpp]: maybe `or to error if a suitable...` would be less dark :)

File: src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp [hunk 2]
```diff
@@ -320,15 +320,20 @@ void Endcap::dxbar_dsigma(
...
-    const std::array<double, 3>& projection_point) const noexcept {
+    const std::array<double, 3>& projection_point,
+    const bool source_is_between_focus_and_target) const noexcept {
   // Try to find lambda_tilde going from target_coords to sphere.
-  // This lambda_tilde should be positive and less than or equal to unity.
-  // If there are two such roots, we choose based on where the points are.
+  // If the target surface is outside the sphere (that is,
+  // source_is_between_focus_and_target is true), then lambda_tilde should be
+  // positive and less than or equal to unity. If the target surface is inside
+  // the sphere, then lambda_tilde should be positive and greater than or equal
```
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp]: would `lambda_tilde must be greater than equal equal to one` by clearer?

**Review by nilsdeppe — DISMISSED** [review]
A couple minor suggestions, please squash if you think they're worth doing

**Review by nilsdeppe — DISMISSED** [review]

**Comment by markscheel** [comment]
Rebased on develop to resolve conflict.

**Review by nilsdeppe — DISMISSED** [review]

File: src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp [hunk 3]
```diff
@@ -436,10 +436,6 @@ class Endcap {
```
Comment (moxcodes) [inline src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:437]: commit message: missing a space in `bothfor`

File: src/Domain/CoordinateMaps/CylindricalEndcap.cpp [hunk 4]
```diff
@@ -110,65 +126,103 @@ CylindricalEndcap::CylindricalEndcap(const std::array<double, 3>& center_one,
...
-  // and the z axis, and let beta be the maximum of that angle (over all points
-  // on the intersection circle).  We demand that beta is less than a certain
-  // maximum opening angle.
-  const double tan_beta = sqrt(square(center_one[0] - proj_center[0]) +
-                               square(center_one[1] - proj_center[1])) /
-                          (max_proj_center_z - proj_center[2]);
-  ASSERT(tan_beta < tan(max_opening_angle),
-         "Opening angle is too large. The map has not "
-         "been tested for this case.");
+    // Check if opening angle is small enough.
```
Comment (moxcodes) [inline src/Domain/CoordinateMaps/CylindricalEndcap.cpp]: It appears that the opening angle condition is only enforced for the case where sphere1 is contained in sphere2. Is that intentional? Why is the condition relaxed for the alternative case?

File: src/Domain/CoordinateMaps/CylindricalEndcap.hpp [hunk 5]
```diff
@@ -62,7 +62,8 @@ namespace domain::CoordinateMaps {
  *
  * ### Restrictions on map parameters.
  *
- * We demand that Sphere 1 is fully contained inside Sphere 2. It is
+ * We demand that either Sphere 1 is fully contained inside Sphere 2, or
+ * that Sphere 2 is fully contained inside Sphere 1. It is
```
Comment (moxcodes) [inline src/Domain/CoordinateMaps/CylindricalEndcap.hpp:73]: On a first approach to this map, it's not necessarily clear what the ramifications are of putting sphere2 inside sphere1 instead. Is it effectively the same map, but the lines drawn from the projection point to the spheres now meet at the points where the plane intersects the outer sphere instead? Is there any reversal of coordinates that the user should be aware of when interchanging sphere 1 and sphere 2 like that?

**Review by moxcodes — COMMENTED** [review]
Looks good, only a couple small comments and suggestions.
I didn't go through and verify all of the numerical choices for bounds on the endcap parameters -- I'll trust that if those aren't working acceptably, the test would at least occasionally fail.

**Comment by markscheel** [comment]
Rebased, fixed typo in commit message, and added a few fixups addressing questions of @moxcodes

File: src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp (hunk 3 above)
Comment (markscheel) [reply src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:437]: fixed

File: src/Domain/CoordinateMaps/CylindricalEndcap.cpp (hunk 4 above)
Comment (markscheel) [reply src/Domain/CoordinateMaps/CylindricalEndcap.cpp]: Yes, it's intentional. I added a comment.

File: src/Domain/CoordinateMaps/CylindricalEndcap.hpp (hunk 5 above)
Comment (markscheel) [reply src/Domain/CoordinateMaps/CylindricalEndcap.hpp:73]: Only Sphere1 has its mapped region bounded by a plane.  To construct Domains out of multiple Blocks, each with their own CylindricalEndcaps, you need touching Blocks to share the same boundaries, which in some cases are the intersection of a sphere and a plane.

I added more text to the dox to try to explain this, using the cylindrical BBH in the 2012 paper as an example.

Comment (markscheel) [reply src/Domain/CoordinateMaps/CylindricalEndcap.hpp:73]: And in all cases you need to worry about reversing/rotating coordinates. CylindricalEndcap is by default oriented in the z direction but in the BBH Domain we will want corresponding Blocks oriented in +x and -x, so in the DomainCreator there will be different OrientationMaps applied to different Blocks.

**Comment by moxcodes** [comment]
looks good, thanks for clarifying those points. I'm happy for you to go ahead and squash.

**Comment by markscheel** [comment]
squashed.

**Review by nilsdeppe — APPROVED** [review]

**Review by moxcodes — APPROVED** [review]

**Pull request merged**